### DISCUSSION TITLE
Welcome to Budibase Discussions!

### DISCUSSION THREAD

**Discussion by joebudi** [Budibase/budibase#615] [Help]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mjashanks** [comment]
Hi, I'm Mike - Co-founder of Budibase!

Feel free to get in touch :)

**Reply by absylver** [reply]
good evening mike

**Reply by mjashanks** [reply]
Hey @absylver 🙂

**Reply by Julian-Dumitrascu** [reply]
Hi! I've subscribed to your [Medium profile](https://medium.com/@mjashanks) and look forward to communicating there with you. I'd like us to communicate publicly using a means that would help us more than GitHub.

**Reply by mjashanks** [reply]
Hi Julian.

GitHub discussions is where our community lives. We will provide support, updates etc here. 

We will not be using Medium for this. Why does medium help you more than GitHub?

**Comment by thewebsitefactorylondon** [comment]
Hello
I'm new to Budibase, could you tell me how to run Budibase on ubuntu server on AWS? Thanks

**Comment by MikeHersee** [comment]
Hi, I downloaded version 0.125 for windows from github, but each time I tried to create a project, after entering the API key and entering the name and an admin user it said, "creating app" (or words to that effect) seemingly indefinitely, while my PC was otherwise sitting idle. So each time I cancelled the application. I tried uninstalling and reinstalling and it didn't make any difference. Each time I could see the app I'd previously created, but was not able to go into any of them. Couldn't work out what I was doing wrong. This was yesterday. Any ideas?

**Reply by mjashanks** [reply]
HI @MikeHersee . That release is marked as pre-release - we did this when we realised that it did not work on windows. Here is the latest release: https://github.com/Budibase/budibase/releases/latest

Unfortunately the latest release has issues when trying to run your web app - but you should still get a good feel for the building experience. This will be fixed in the next few days.

In general, we are in private Beta at the moment. Alot of things are changing right now - but we are nearly there, and things will settle down soon. We plan for a stable beta by the end of next week.

Thanks for your patience!

As always, feel free to come and chat to us on Discord, or post in here of course :)

**Reply by MikeHersee** [reply]
Ok, thanks for that. I'll probably wait a few days then, but it sounds interesting.

**Comment by MikeHersee** [comment]
Database relationships:

Hi, I've downloaded version 0.2.0. When creating files, how do you set which keys or combinations of keys must be unique? I didn't see a feature for that. It's important for building proper relational databases. And also, how do you define foreign keys.

**Comment by MikeHersee** [comment]
Ok, I've found out how to do foreign keys, but defining the unique key field or field combination I can't see

**Reply by shogunpurple** [reply]
@MikeHersee unique columns or composite columns are not something that is supported at the moment. This isn’t to say they won’t be in future but it’s not on our immediate roadmap. Please feel free to raise a discussion thread around this and we can discuss further.

**Comment by MikeHersee** [comment]
Foreign keys and record descriptions

I'm not sure what your intended direction is, but the focus seems to be on not having to create code to do things. This means that the database needs to do things such as integrity checking, so that setting a field or combination of fields to 'unique' means that any attempted violation of that will be automatically handled and won't have to be specifically coded for.

Another issue is that it is normal in reference files to have a unique key for each entry that does not necessarily have meaning in and of itself, but each record contains a description of that record. For example, you might use 'vehicle registration' as the unique key for each vehicle, Defect types might have codes, such as '0001', which could be 'brakes pulling to one side', '0002' which could be 'broken window', '0003', which could be 'indicator light out', and so on. It would be the code you would want to insert into the 'defect' record as the foreign key, but when you're looking up the defect type to insert into the instance of a defect, you want the description field displayed rather than the meaningless code. I haven't found a way to do this yet.

**Reply by shogunpurple** [reply]
@MikeHersee please create another thread for this - this is the welcome thread, not a thread for technical discussion. 

To address your points:

**Unique Constraints**
Composite keys are probably not something we will introduce with budibase in the short term for comprehension reasons. However the concept of unique records, or preventing someone from creating a record with a duplicate column value is something that is common in most software. Do you feel like marking a column as "unique" is a nice way to achieve this from your perspective? Having unique columns is a bit more flexible than just designating a single unique, primary key for a table.

**Your Example**
You can actually model your example right now with budibase, but it's a little strange (We are addressing some small UX issues this week)

You need to choose a "Primary Display Column" for each of your tables. The primary display column is essentially what gets displayed when you view a relationship. So in order to model your example:

- Create a "Vehicle" table and set primary display column to Registration
- Create a "Defect Type" table with the code (0001) and description columns (brakes pulling to one side). Set the primary display column to "Description"
- Create a "Defects" table with 2 relationship columns - to vehicle and to defect type. This means you can now add a defect, and select your vehicle by reg and defect type by description.

![Screenshot 2020-10-12 at 10 40 59](https://user-images.githubusercontent.com/11256663/95731430-6e884d00-0c77-11eb-9365-82850eb837fa.png)
![Screenshot 2020-10-12 at 10 30 23](https://user-images.githubusercontent.com/11256663/95730640-73003600-0c76-11eb-8c21-5e6b76459e26.png)
![Screenshot 2020-10-12 at 10 29 37](https://user-images.githubusercontent.com/11256663/95730644-74c9f980-0c76-11eb-9c11-5296df6a11ff.png)

Hope this helps! Your feedback is super valuable to us - so thank you for taking the time to write this up and use budibase. We'd love to continue this discussion on the other thread.

**Comment by damienvc** [comment]
Hi all. I thought I'd introduce myself as a new user (and also new to programming although I used to code when I was a lot younger!). I've only just downloaded Budibase but I'm really liking it so far! So, hello.

**Reply by mjashanks** [reply]
Hi Damien! Welcome to Budibase :)

Its great to have you, please feel free to raise any issues, discussion - or come and chat to us in our [Discord](https://discord.com/invite/rCYayfe). We'd love to know more about what you like, and what we can improve upon - especially as someone with a little coding experience.

**Comment by Caterina1994** [comment]
Hello!
Im having problems to connect with AIRTABLE.
I put my API key and it seem to work fine in zapier, but doesnt recognize it in BUDIBASE

**Reply by joebudi** [reply]
Hey.

This may be the possible solution which will be released on Thurs:

https://github.com/Budibase/budibase/issues/1027

**Reply by Caterina1994** [reply]
hEY @joebudi ..My problem is that i cant even connect to airtable. 
I put my API KEY and it wont let me create.
The create button is desactivated

**Reply by shogunpurple** [reply]
Hi @Caterina1994, please make sure that your airtable connection is set up correctly:

- The apiKey is your airtable API key
- The base is your **base id**, not your base name. The baseId can be found here https://airtable.com/api

Once you have inputted those values, are you still unable to connect to airtable?

> The create button is desactivated

The query save button will be deactivated until you run the query first to test it. This is not possible because your connection to airtable is not working.

**Reply by Caterina1994** [reply]
Hey @shogunpurple  it works! It was a stupid error . I didnt name my Base so the CREATE button was disabled. :(.
Now i how to figure out the Query part.

**Comment by pierrelandoin** [comment]
Hi! I am a French entrepreneur, with a strong interest in no-code / low-code technologies. I am happy to discover Budibase. Could anyone help me answer these 3 preliminary questions:
1- When creating an app on Budibase, can I export the source code? 
2- If yes, is this source code human-readable and easily modifiable?
3- If I have modified manually the source code of my app, can I reinject it into Budibase to keep using the platform's tools?

**Reply by joebudi** [reply]
Hey, and welcome. Budibase does export your app but it's not possible to code against it. This is something we want to make possible in the future.

**Comment by waispirzad** [comment]
Hi, my name is Wais Pirzad and I am a business analyst. I am very excited about Budibase. I have played with it for a few hours and have created more than I could create with any other low code solution. 

Why do I try to Budibase?
Because I like to create internal tools as efficient as possible. There are tools such as Mendix and Bubble I consider and experiment with. The problem is that both are easy but complex in nature. So it takes me a lot of time to build internal apps. Why because I have to design it or buy change a design  of a template. This is easy but takes a lot of time. Both data and automation functions are great. But their design features are time consuming. And the combination of the data, design and workflows makes development complex. Because al three has to be constantly changed in a complex way to make a change.

So what did I like of Budibase?
- I didn't have to spend time on design because it is an internal tool. This is already solved. So it makes me spend les time to reach my goal. +1 for Budibase. 
- It  is intuitive in such a way that it enabled me to create an basic app to maintain projects and understand some the functionalities without reading the documentation. +1 Budibase.
- it is easy to structure and link tables
- I love the ease of testing locally. Helpes me to accelerate my development.

I will proceed with experimentation. And will provide feedback.

My goal is learn to develop with a low code tool so I can not only build quickly internal tools but also startup ideas I have. So let's see if Budibase can be my solution. 

Keep up the good work. 👍👍👍

Wais

**Reply by mjashanks** [reply]
Thank you Wais! Feel free to reach out if you have any questions.

**Comment by davepile** [comment]
Hi,

I am trying out Budibase by running through the tutorial (https://docs.budibase.com/tutorial/5.-designing-your-app). The app seems to have changed quite a bit since the tutorial was made. I am managing to find my way except for one step.

In tutorial Design step #5 where we create a stacked list for Deals I cannot add a link to the Deals Detail (edit) page. In the tutorial text it says "DestinationUrl, choose Deals Details" and in the video the Link URL field is shown as a combo box where you can select the "Deals Detail" route. 
However in my Budibase builder, the Link URL field is a free text field. I have tried entering various values such as 
    /deals/:id
    /deals/{{ New Repeater.Deals.Auto ID }}
    /deals/{{ New Repeater.Deals._id }}
etc. with no luck
Anyone know how this url should be structured?

Other than that, so far it looks really good and I am looking forward to seeing how I could use Budibase for small in house apps
Thanks

**Reply by davepile** [reply]
After a bit of trial and error I had some success with this - 
Link URL:      #/deals/{{ New Repeater.Deals._id }}
 
Happy to have some feedback if this is an acceptable way to do this.

**Reply by mjashanks** [reply]
Hi @davepile . This is indeed a bug. The URL that you have guessed above is correct - i.e. the `#` should be there. The Card components should know to put this in automatically. Thanks for reporting. I logged it here  https://github.com/Budibase/budibase/issues/1358

**Comment by dand31** [comment]
Hi, 

Is it possible to upload videos (mp4 etc) to your app? I want the users to be able to upload content such as documentation, videos, and even film videos within the app. Is this possible?

Thanks,
Dan

**Reply by joebudi** [reply]
Hi @dand31 . Yes, it is possible to add videos as attachments.

Here is a quick video to demonstrate:
https://www.loom.com/share/dd6532085dab412cb5805757de18f110
I hope this helps.

**Comment by LugarJE** [comment]
Hi, I'm grateful to find this (just before folding up). Been hitting new skills hard since last July 2020. This past July, I have been installing Postgres, K8s, minikube (k2s?), KVM and learning micro services. Using a HP laptop with Debian Buster. I really, really need to begin an alpha web app in a month! Keeping it simple (microservices) for 3 levels of user-- fan, supporter and participant. I need help with re-installation of KVM(2). Want/need to use Laptop for local dev. Want to be a startup success story for this platform.  My background: EE since 1992, #IoT dabbler on and off, Former MS Access RAD, developer by need not official training. Really want to deliver this simple piece by Veterans day! (and test it sooner.

**Comment by Rickwright12** [comment]
Hi All,  new to this with low skill level.  :(

**Comment by michaelajiri** [comment]
Hello everyone. feels good joining this platform. Please is there a way I can connect to an online database like MySQL?

**Reply by gerhardmpl** [reply]
Yes, you can connect to MySQL and there is an [example in the documentation](https://docs.budibase.com/quickstart-tutorials/crud-app-with-sql/connecting-to-your-sql-tables) to give you some ideas. One you have the Datasource configuration right, you can create tables and screens for MySQL.

**Comment by dan-lindley** [comment]
Hi, I'm Dan 

A complete Budibase noob. I've dabbled years ago with access and never really got into the web side of things. Anyhow, the time has come and I want to create an app. My first thought was with access and then I had a look around and came across budibase. 

I do love the layout in the backend. It feels familiar. I'm just starting to read around the community pages and I don't think what I wish to do will be possible with budibase ;( - I want to create a charitable application for users to register, access and manage their own collections of data (this would be the same datasets among all collections) and i've seen this is likely SaaS (??? another new term for me). It is a shame as i really like the feel of the platform and it seems less cluttered than others i've seen.

I do hope that there is a possible user signup/group signup for the future as this would be a good platform to build in. it feels so intuitive.

Good luck guys and keep up the excellent work. I'll stick around just for playing with!

Dan

**Reply by mjashanks** [reply]
Hi Dan! Welcome :)

This month, we are adding support for the Budibase REST API. This means that you should be able to code your own signup process.

Self-signup is not on our roadmap, but I wouldn't rule it out eventually.

**Reply by dan-lindley** [reply]
That's welcome news. I can't find anything quite like your platform and I've dabbled with so much. Having a registration process and a user management system structured within would be very handy. Looks like I'll keep dabbling with what I have built so far!
Thank you.

**Comment by ThoBode** [comment]
Hi, I am new to budibase. I like the project very much. However, for my application, I need access protection at the row level of a table. 

Is it possible to get the name or ID of the currently logged in user and use it in queries?

Thanks, Thomas

**Reply by mjashanks** [reply]
Hi @ThoBode - welcome :)

> I need access protection at the row level of a table.

This is not possible at the moment. As you say, you can pass in the user ID to a custom query. However, the ID gets passed from the frontend, so this does not secure the API against users passing any ID in.

I would love to see this as a future feature though, but it's not on our roadmap at the moment.

> Is it possible to get the name or ID of the currently logged in user and use it in queries?

Yes - its possible the use the current user in queries. You will need to:
1. Setup your query
2. Add bindings to your query
3. In the `Design` section, pass values into your bindings. Here, you can use properties like `{{ Current User.email }}` (or other user fields)

**Reply by ThoBode** [reply]
Thank you for the quick feedback. I will try this.

Since it's an internal application, the security requirements are not so high.

Thomas

**Comment by ShivaaneeSharma** [comment]
Hello
I'm new to Budibase, I'm facing issues while creating the app. It says error while creating the app when I select the create app icon and write a name and then click the create click button

![Screenshot 2022-03-21 115814](https://user-images.githubusercontent.com/102014698/159228325-17942e3a-aa7e-4096-82d7-d4ca295d89c4.jpg)

**Reply by shogunpurple** [reply]
Hi @ShivaaneeSharma - I can't seem to recreate this one. Are you using budibase cloud? If so could you let us know your tenant name and we will investigate. 

It would also be useful to see the output of your browser console. You can do this by right clicking and choosing "Inspect element", then going to the "Console" tab.

Thanks!

**Comment by ShivaaneeSharma** [comment]
@shogunpurple, Thank you so much for answering to my query. This shows when I go to console. 
May I know what a tenant name is?
Also, I'm not able to create an app even by using a template.

![Screenshot 2022-03-21 144154](https://user-images.githubusercontent.com/102014698/159232259-e8418410-62c2-4408-9108-6fa14023a510.jpg)

**Reply by shogunpurple** [reply]
Hi @ShivaaneeSharma - your tenant Id is the first part of your budibase URL. `yourtenantname.budibase.app`. If you share that, I should be able to check the logs to see what happened. Thanks!

**Reply by shogunpurple** [reply]
Hi @ShivaaneeSharma - I looked into this and it appears that the document update conflict seems to be happening on the usageQuota document. This conflict appears to have been resolved, so you should be able to create an app now. Can you try again? If not I'll continue debugging. Thank you

**Reply by ShivaaneeSharma** [reply]
> 

@shogunpurple I tried again and I am still not able to create an app. Sorry for the late response.
Thank you so much for looking into this issue.

**Comment by ShivaaneeSharma** [comment]
https://shivaaneefirstapp.budibase.app/builder/portal/apps
Thanks!

**Comment by radu-tataru** [comment]
Hi. I really like budibase and i am trying to build a small application with it, as a proof of concept for a future client (an auto repair shop). I am now struggling to automate an SMS notification system. These SMSs are reminders for the shop apointments.
My approach is to set a cron trigger that runs every morning at 8 AM, reads all the rows from the appointments table that have the appintment date = {{ date now }} and for each row i generate a webhook for zapier (SMS.to) with the phone number, client name etc as payload.
My question is: how do i create a webhook for each row (is this possible or my whole approach is wrong?). If my approach is wrong how should i try to accomplish this? 
Thank you!

**Comment by abhishikthchandaka** [comment]
Hello @shogunpurple, I've been experimenting with BudiBase to start using it as internal admin portals for multiple products within our company. Everything has been great so far, love the UI. 

However, there has been one major point of concern that has blocked me from moving forward with BudiBase's adoption fully - setting a column as unique. This seems to me as something that's a very basic requirement for any CRUD solution such as BudiBase, so I think I might be missing something here. I saw you answer a similar question from @MikeHersee on 10 & 12 Oct 2020, but I couldn't find anything beyond that.

Could you please help me out with this? Is it already implemented? If so, how can I use it and if not, is there any common work around that people use / is there any update on this being implemented in the near future (I couldn't find any git issues related as well). Thanks :)

**Comment by preethamts** [comment]
hi @shogunpurple i created a table called order request , i want to create an instance for the same, i dont see one to one relationship while creating the table.

**Comment by geoffgpe** [comment]
Created an App and have connected to a substantially sized existing MySQL Database for the Datasource. I am having troubles creating a Table and would like to disconnect from this Database and try a newly created database with no tables and build up from there.

If I delete the MySQL Datasource (the substantially sized existing MySQL Database) in the App, just confirming it does not delete the Tables in the Database on the Server.
